### ISSUE TITLE
Photon: add option to flush the cache for the whole site

### ISSUE THREAD

**Issue by nengineer** [Automattic/jetpack#917]
You really need a way to clear the Photon cache for a site from within JetPack.

**Comment by maivutan** [comment]
Any updates?

**Comment by jeherve** [comment]
@maivutan We'll post here as soon as we make progress on this.

**Comment by rdreessen** [comment]
Any updates on this? Considering adding to a large site. Thanks!

**Comment by kraftbj** [comment]
@rdreessen As stated above, we'll post here when there are updates. Thanks!

**Comment by jeherve** [comment]
Also requested here:
https://wordpress.org/support/topic/jetpack-and-ewww-image-optimizer-images-problem?replies=6&view=all#post-7335364

**Comment by samhotchkiss** [comment]
Asked our systems team about the technical limitations around this so that we can decide how to handle

**Comment by samhotchkiss** [comment]
Based on the technical limitations of our CDN, there isn't a way to do this on a site-wide basis at this time.

**Comment by nengineer** [comment]
So, it has taken you over a year to figure this out?

**Comment by maivutan** [comment]
Why you can do this manually but the site-wide?

**Comment by kraftbj** [comment]
> So, it has taken you over a year to figure this out?

We left it open in hope the technical limitations would change, but we've opted to close the issue at this time until such a time the infrastructure would make this feasible. In short, it didn't take us a year to determine the limitation, but a year to decide to not leave it open pending something outside our team's scope to change.

> Why you can do this manually but the site-wide?
Manually, we're able to clear out specific individual URLs. We're unable to do this en mass and it's pretty time consuming, unfortunately.

**Comment by rothschild86** [comment]
Ok, so not site-wide. How about a button for each individual image in the Media library?

**Comment by jeherve** [comment]
Another vote for a user-facing tool to flush Photon's cache for specific images:
https://wordpress.org/support/topic/option-to-clear-image-cache?replies=1&view=all

**Comment by jeherve** [comment]
Also requested here:
https://wordpress.org/support/topic/please-remove-all-cached-images?replies=1&view=all

**Comment by rothschild86** [comment]
As an additional note - the goal is to make photon work with plugins such as https://wordpress.org/plugins/enable-media-replace/

As of now, an old image is being served after it has been replaced in the media library.

one idea:
adding a date-modified-time-stamp to the end of photon url would tell it that the image has changed and needs to be refetched...

**Comment by sneader** [comment]
Yet another vote for this feature. I have to disable Photon on yet another site, because the customer can't resize a previously uploaded image.  Photon is so great, but made nearly unusable due to this inability to flush the current images from the CDN.

**Comment by kraftbj** [comment]
@sneader If they're resizing the image within wp-admin, that should generate a new image that is seen as a new image by Photon. For resizing an existing image, then yes, renaming the image is the way to go at this time.

**Comment by tonnybukdahl** [comment]
Where would one have to turn to, to have the images uploaded by Photon removed permanently?

I'm asking since I just recently found out about this issue, and being a freelance photographer shooting copyrighted material, I'd like to be the only one hosting my photos

**Comment by jeherve** [comment]
@tonnybukdahl You can contact us via [this contact form](http://jetpack.com/contact-support/), and we can flush the cache for any image URL you send us.

Thanks!

**Comment by MKRhere** [comment]
It's been 2 years since this issue was raised and it's going to be a 3rd.. As much as Photon is great, it really needs this option. One more vote.

**Comment by laltroweb** [comment]
ok... im not alone with this: https://wordpress.org/support/topic/maintain-querystring-with-photon-image/#post-8536649 :/

**Comment by sterndata** [comment]
I've changed themes and images now look "thready" because Photon is serving them with the wrong size.  The only way to make my site look OK is to disable photon.  It would be nice to be able to clear cache.

**Comment by jeherve** [comment]
@sterndata Clearing the cache wouldn't help in your case I'm afraid. When a new image size is needed (like when you switch themes), a new image is built from scratch in Photon. The old cached images aren't used anymore.

Could you send us an email via [this contact form[(http://jetpack.com/contact-support/) so we can take a closer look? It might also help if you could let us know where we can get a copy of the theme you're using, so we can run some tests on our end as well.

Thanks!

**Comment by sterndata** [comment]
The theme is twentyseventeen, The image is the dreamweaver affiliate banner on the home page of www.sterndata.ccom

**Comment by jeherve** [comment]
@sterndata The images don't seem to use Photon anymore, but I don't see much difference when loading the original image vs. the one that's served via Photon:
https://www.sterndata.com/wp-content/uploads/DH_Banner_728x90.gif
https://i0.wp.com/www.sterndata.com/wp-content/uploads/DH_Banner_728x90.gif

Do you still experience issues when Photon is enabled?

**Comment by sterndata** [comment]
Without Photon
![image](https://cloud.githubusercontent.com/assets/1135810/22895215/4ea91956-f1e1-11e6-84f5-e3872d1a063d.png)

With Photon
![image](https://cloud.githubusercontent.com/assets/1135810/22895191/3d148ff4-f1e1-11e6-92dd-7fee133f0ab0.png)

You can see that the "with photon" image is reduced in size (compare to the text above)

**Comment by jeherve** [comment]
Could you post the content as it appears in your post editor, in the HTML tab, so I can try to reproduce?

Thanks!

**Comment by sterndata** [comment]
`Click on the images below for more information:

<a href=" http://www.dreamhost.com/r.cgi?2167333"><img class="size-full wp-image-1632 aligncenter" src="https://www.sterndata.com/wp-content/uploads/DH_Banner_728x90.gif" alt="" width="728" height="90" /></a>

<a href="https://www.digitalocean.com/?refcode=330db2db9b21"><img class="aligncenter wp-image-1222 size-full" src="https://www.sterndata.com/wp-content/uploads/ssd-virtual-servers-banner-320x50.jpg" alt="ssd-virtual-servers-banner-320x50" width="320" height="50" /></a>`

**Comment by jeherve** [comment]
Thanks!

This is odd. On your site, the HTML we get when Photon is enabled is as follows:

```html
<p><a href=" http://www.dreamhost.com/r.cgi?2167333"><img data-attachment-id="1632" data-permalink="https://www.sterndata.com/website/dh_banner_728x90/" data-orig-file="https://i2.wp.com/www.sterndata.com/wp-content/uploads/DH_Banner_728x90.gif?fit=728%2C90&amp;ssl=1" data-orig-size="728,90" data-comments-opened="0" data-image-meta="{&quot;aperture&quot;:&quot;0&quot;,&quot;credit&quot;:&quot;&quot;,&quot;camera&quot;:&quot;&quot;,&quot;caption&quot;:&quot;&quot;,&quot;created_timestamp&quot;:&quot;0&quot;,&quot;copyright&quot;:&quot;&quot;,&quot;focal_length&quot;:&quot;0&quot;,&quot;iso&quot;:&quot;0&quot;,&quot;shutter_speed&quot;:&quot;0&quot;,&quot;title&quot;:&quot;&quot;,&quot;orientation&quot;:&quot;0&quot;}" data-image-title="DH_Banner_728x90" data-image-description="" data-medium-file="https://i2.wp.com/www.sterndata.com/wp-content/uploads/DH_Banner_728x90.gif?fit=300%2C37&amp;ssl=1" data-large-file="https://i2.wp.com/www.sterndata.com/wp-content/uploads/DH_Banner_728x90.gif?fit=644%2C80&amp;ssl=1" class="size-full wp-image-1632 aligncenter" src="https://i2.wp.com/www.sterndata.com/wp-content/uploads/DH_Banner_728x90.gif?resize=644%2C80&amp;ssl=1" alt="" width="644" height="80"></a></p>
```

On my end, here is what I get:

```html
<p><a href=" http://www.dreamhost.com/r.cgi?2167333"><img class="size-full wp-image-1632 aligncenter" src="https://i2.wp.com/www.sterndata.com/wp-content/uploads/DH_Banner_728x90.gif?resize=644%2C80&amp;ssl=1" alt="" width="644" height="80"></a></p>
```

That's most likely because the image is not attached to that page on my site, so it can't generate additional `srcset` parameters.

Still, the `width` and `height` parameters aren't correct.

cc @kraftbj Any idea where these ` width="644" height="80"` parameters are coming from?

**Comment by kraftbj** [comment]
https://github.com/WordPress/WordPress/blob/1015c3b3a07ad6b530bb3aed7e0e11549da791cb/wp-content/themes/twentyseventeen/functions.php#L236

There are different content widths on Twenty Seventeen if content is on the front page vs a regular page.

**Comment by jeherve** [comment]
@kraftbj Thanks!

@sterndata If you'd like, you could use the code snippet [here](https://github.com/Automattic/jetpack/issues/4175#issuecomment-227282032) to force Photon images to be larger than the theme's `$content_width` value.

**Comment by sterndata** [comment]
Thank you very much!

**Comment by rambabusaravanan** [comment]
Is the "Clear Cache" option enabled?

A heavy uncompressed image was uploaded and photon was enabled and so the image was cached
Now I compressed the image to a half size and photon is still having my old heavy image.

When photon enabled, heavy images reduces my page speed score.
When photon disabled, my page have a good speed score due to lightweight images.

Please give us provision to clear cache for Photon images.

**Comment by jeherve** [comment]
Also in 913172-zen

**Comment by emdeedee** [comment]
Is there any progress on this at all? This effectively means there is no way to edit an image without renaming it.
Started using "Image Regenerate & Select Crop" to tweak page images etc. and none of these get to photon cache which is really frustrating.  Having to turn off photon :-(

**Comment by Sakaane** [comment]
Adding another vote. It's mind-boggling to think this module was released and has gone this long without some level of purging or refresh functionality, and that users are not forewarned about this limitation before they enable Photon. Images are cached "forever"?? The storage requirements must be enormous. Someone actually thought that was a good idea??

There are currently almost 750 images on my website. I now have a need to update all but a small handful. It wasn't until I got started that I figured out Photon wasn't picking up the changes. I expected that turning the module off and then back on again would be enough to force Photon to see my site as "new" and overwrite the previous files, but nope.

Expecting me to change the filenames on over 700 images (and then having to edit all the corresponding posts and pages those images were previously linked to) OR having to identify and submit 700+ URLs via a contact form is ludicrous. I don't want to do that—do YOU want me to do that, _really??_

Since the originating site's root domain is present in the Photon URLs you'd think that would be enough to work with, particularly when a user needs a full purge. If Photon is smart enough to know which image to serve back to the user's domain in place of the original, then its unique URL and the association to the original has to be stored somewhere. You guys don't know how to write queries for specific character strings?

Additionally, since Photon is smart enough not to get images mixed up when serving them back to the originating site, it should also be smart enough to link those unique URLs to the original image in the Media Library. If JetPack support can manually remove the Photon image by knowing what the URL is, then a button that does exactly the same thing can surely be added to the Media Library.

I'm forced to turn Photon off until more user-friendly functionality is added. I'm not very happy about this, not so much about any potential loss in loading speed for my website, but more that it means I have absolutely no control over _my_ content already captured by Photon unless I do the aforementioned 700+ URL slog. If I had known about that when I first installed JetPack I never would have turned Photon on at all.

**Comment by sneader** [comment]
@Sakaane I don't think you could have stated the problem any better. Thank you. I just hope someone at Automattic is listening.

**Comment by jeherve** [comment]
Thanks for the feedback. I definitely get your frustration. Photon's behaviour is not straightforward here, and gets in the way of your work without warning.

Let me try to address some of your questions:

> The storage requirements must be enormous. 

They are indeed. That is what Photon was built for.

> There are currently almost 750 images on my website. I now have a need to update all but a small handful. 

Could you give me more details about your process here? WordPress itself stores meta data about each one of your images, and creates resized versions of each one of your images as you upload them to your Media Library. When you update 750 images at once, do you also update the resized versions? How do you go about doing that?

> You guys don't know how to write queries for specific character strings?

Unfortunately the solution is not straightforward. The flushing mechanism has to apply to all the servers where the cached images are stored; ideally we could flush cache for all images using your domain name as a wildcard for example. Unfortunately that is not something that is easy to implement on our servers.

> since Photon is smart enough not to get images mixed up when serving them back to the originating site, it should also be smart enough to link those unique URLs to the original image in the Media Library.

That would not be very helpful unfortunately. Most of the time you would not want to flush cache for only the original version of an image; you would want to flush cache for all resized versions of that image to refresh that image. To give you an example, here is an image as it lives in my Media Library:
`https://jeremy.hu/wp-content/uploads/watson-5818.jpg`

Flushing `https://i0.wp.com/jeremy.hu/wp-content/uploads/watson-5818.jpg` is pretty useless to me; I would want all resized copies to be flushed, i.e. `https://i0.wp.com/jeremy.hu/wp-content/uploads/watson-5818.jpg?w=365&h=244`. There can be a very large number of variations depending on how and where the image is used. 

As a result, a button in the Media library would have to be able to catch all variations of the image at once. And we're back to our problem of wildcards. :(

There may be a solution to that problem, but we haven't developed it yet. That does not mean we never will, but until then I'm afraid Photon just isn't quite the tool one can use when making changes to existing images outside of the WordPress media library.

I hope this clarifies things a bit. 

When we do find a solution, we will definitely update this issue, so each one of the people on this thread will be notified!

**Comment by Sakaane** [comment]
> and creates resized versions of each one of your images as you upload them to your Media Library. 

My images are already sized to the exact dimensions I want to display them in. I delete all the alternate ones WordPress creates because those dimensions are useless to me—except for the very smallest thumbnail image, but only to avoid getting a flood of 404 errors when loading the Media Library since that's the only place those thumbnails are used. (There needs to be an off button for that image generation "feature". I don't need WordPress to babysit me by flooding my server with files I don't require. No, it really isn't a problem that the alternate images don't exist. Really.)

> How do you go about doing that?

Photoshop > Overwrite old image file on local HDD > FTP to server.

When the replacement image has the same dimensions as the old one, there's nothing else to do after FTPing the new file up. I never upload images using the dashboard. I prefer my uploads folder to be organized a certain way, so I maintain it myself directly.

If the replacement image has different dimensions, I have to take the extra step of deleting the existing Media Library entry before FTPing the replacement image, then using the Add From Server plugin to get the new image into the Media Library, as the Media Library doesn't pick up the new dimensions automatically. Using Add From Server triggers the alternate versions to be generated so I use FTP to delete them. In my case, replacing with different dimensions doesn't happen very often though and is not applicable to the 700+ images I'm working on right now.

Brand new images are uploaded via FTP as well, added to the Media Library via the Add From Server plugin, and then the excess alternates are deleted.

> Unfortunately that is not something that is easy to implement on our servers.

I find that hard to believe. A server is just a particular kind of computer that stores a bunch of data until it needs to be retrieved. Since the Dark Ages of computing people have been able to key searches with wildcards to find the files they need. I'll grant you that nowadays computer systems are much more complex but the fact remains that it's still all 0s and 1s and a computer can be made to do a wildcard search, whether it's looking on one drive or many, one machine or a network of them. It doesn't have to be pretty or slick or fancy; a KISS formula that works in a timely manner is all that's needed. If your servers are actually set up in such a way that you really truly can't do something a simple as a wildcard search, trash your setup. Other CDNs can flush cache via integrated user side requests so it must not be that hard.

> you would want to flush cache for all resized versions of that image to refresh that image.

Yes, of course. This doesn't change the fact that for a full site purge all that is required is the originating domain name. In such a case the names of the files, or the specific dimensions tacked onto the alternate versions of those files (regardless of how many versions there are), are irrelevant. For someone who only needs specific files flushed, the combination of the domain name and the specific filenames (without the dimension modifiers) would be enough.

WordPress is smart enough to know that it needs to delete all alternate versions of a file when the original version is deleted in the Media Library—from one little link that says "Delete Permanently". JetPack should be able to piggyback on that to facilitate deleting files from Photon. JetPack should also be able to apply similar theory to a link or button that says "Regenerate Photon file" (or "files" if you prefer).

> There may be a solution to that problem, but we haven't developed it yet. 

Don't reinvent the wheel. The solutions already exist.

PS: Be transparent. Warn JetPack users up front, right there below the Photon module's toggle, what they are getting themselves into and what the ramifications are if they turn Photon on even once.

PPS: Be extra transparent. Warn JetPack users up front, right there below the other applicable toggles, when some other module (such as Tiled Galleries) requires Photon to work and what the ramifications will be, even when the actual Photon module itself is turned off. Don't be sneaky like that.

**Comment by jeherve** [comment]
Thanks for the extra details.

> There needs to be an off button for that image generation "feature". I don't need WordPress to babysit me by flooding my server with files I don't require. No, it really isn't a problem that the alternate images don't exist. Really.

There are plugins that can help you with that ([here is an example](https://wordpress.org/plugins/image-sizes/)), but I'm not sure removing them is ever a good idea. Yes, you do save space on your server and that's important if you have limited disk space there, but those alternate image sizes are also important, and can be used by WordPress itself everywhere on your site when your site's visitors access your site from mobile devices or smaller devices. Since WordPress 4.4, WordPress references these alternate image sizes thanks to additional attributes and that allows smaller devices to load them and thus avoid loading large images when they don't need them. You can read more about it here:
https://make.wordpress.org/core/2015/11/10/responsive-images-in-wordpress-4-4/

That said, I do not mean to question your setup; if that workflow works for you you should be able to use it, and Photon should not get in the way. When we'll have managed to develop a solution that works, we'll update this issue accordingly and will mention it in the changelog.

**Comment by mzakariya** [comment]
972525-zen

Photon user is looking to completely nuke the cache, but we do not offer that functionality. Reporting here.

**Comment by prodancer2015** [comment]
This Photon issue with no cache control and the constant fight to get images changed has been ongoing for 3yrs???

I understand that coding and building PHP for WordPress takes time, but really, this is far too long to resolve an issue that is clearly widespread. 

This feature of JetPack has been included for a long time and no change to how it works or how it creates headaches for website / media management. The only change I experienced in this was the name from Photon to "Image Performance". 

Here is another nail in the coffin of this plugin. 

I had to turn off "image performance" feature because I could not get the replacement images to show. Once I turned it off, the replacement images showed. For days I fought with this caching feature to change the image but all attempts failed.

What does this say for live business websites that change product content? 

It says they cannot conduct business because this caching of images, where no control over the cache exists, business can never show new products or replacement product images. 

To show what I mean here, I shut off the feature a few days ago to get the replacement images to show. Today, I reactivated the "image performance" feature of Jetpack and guess what happened? - the old images replaced the new ones. - really!

One would think that this user-unfriendly feature would regenerate its cache when turned off and back on, but No, it shows the old images no matter what and no way to resolve it.

Tell you what, since this issue has never been addressed and never fixed for over 3 yrs, just remove the feature and don't bother adding it back. Keeping this unworkable feature only screws with people's websites, to the ones who are unaware of this serious issue.
